packages/synapse-react-client/src/lib/utils/hooks/SynapseAPI/KeyFactory.ts

```diff
@@ -0,0 +1,562 @@
+    tableQueryBundleRequest: queryBundleRequest,
+  }),
+
+  fullTableQueryResult: (queryBundleRequest: QueryBundleRequest) => ({
+    ...entityQueryKeyObjects.entity(queryBundleRequest.entityId),
+    scope: 'fullTableQueryResult',
+    tableQueryBundleRequest: queryBundleRequest,
+  }),
+
+  boundJSONSchema: (id: string) => ({
+    ...entityQueryKeyObjects.entity(id),
+    scope: 'boundJSONSchema',
+  }),
+
+  schemaValidationResults: (id: string) => ({
+    ...entityQueryKeyObjects.entity(id),
+    scope: 'schemaValidationResults',
+  }),
+  header: (id: string) => ({
+    ...entityQueryKeyObjects.entity(id),
+    scope: 'entityHeaders',
+  }),
+  headers: (references: ReferenceList) => ({
+    ...entityQueryKeyObjects.all,
+    scope: 'entityHeaders',
+    entityHeadersRequest: references,
+  }),
+}
+
+const downloadListQueryKeys = {
+  /* Key used to invalidate all download list queries */
+  base: 'downloadList',
+  availableFiles: (request: AvailableFilesRequest) => [
+    downloadListQueryKeys.base,
+    'availableFiles',
+    request,
+  ],
+  availableFilesInfinite: (request: AvailableFilesRequest) => [
+    downloadListQueryKeys.availableFiles(request),
+    'infinite',
+  ],
+  getActionsRequired: () => [downloadListQueryKeys.base, 'actionsRequired'],
+  getActionsRequiredInfinite: () => [
+    ...downloadListQueryKeys.getActionsRequired(),
+    'infinite',
+  ],
+  getStatistics: () => [downloadListQueryKeys.base, 'statistics'],
+}
+
+function removeTrailingUndefinedElements(params: any[]): any[] {
+  const result = [...params]
+  let lastElement = result[result.length - 1]
+  while (
+    lastElement === undefined ||
+    // Also remove empty objects or objects where only keys with undefined values are present
+    (typeof lastElement === 'object' &&
+      isEmpty(pickBy(lastElement, v => v !== undefined)))
+  ) {
+    result.pop()
+    lastElement = result[result.length - 1]
+  }
+  return result
+}
+
+const ACCESS_REQUIREMENT_QUERY_KEY = 'accessRequirement'
+
+/**
+ * Returns a react-query Query Key.
+ *
+ * Co-location of react-query keys for queries related to Synapse data. Two objectives of this strategy are
+ * - minimize the risk of inappropriate cache key collisions
+ * - simplify clearing the approprate query caches when the entity data is updated
+ *
+ * For more information, see https://tkdodo.eu/blog/leveraging-the-query-function-context
+ */
+export class KeyFactory {
+  accessToken: string | undefined = undefined
+  constructor(accessToken: string | undefined) {
+    this.accessToken = accessToken
+  }
+
+  /**
+   * Returns a react-query Query Key. Prepends the key with the access token, so that the key is unique per-session.
+   * Also removes trailing undefined elements from the passed args, so a key can be generated for query invalidation.
+   * @param args
+   * @private
+   */
+  private getKey(...args: any[]): QueryKey {
+    return [this.accessToken, ...removeTrailingUndefinedElements(args)]
+  }
```

Comment:
Every function just uses this. Remove trailing undefined elements for better key matching